### PR TITLE
[JENKINS-37625] Don’t get the request.remoteUser if it’s a deferred authentication

### PR DESCRIPTION
**Pull request by cyrille-leclerc**
[JENKINS-37625] Don’t get the request.remoteUser if it’s a deferred authentication (`org.eclipse.jetty.server.Authentication.Deferred`) to prevent "IllegalStateException: Response is committed"

```
java.lang.IllegalStateException: Response is committed
	at org.eclipse.jetty.server.Request.getSession(Request.java:1400)
	at org.eclipse.jetty.security.authentication.FormAuthenticator.validateRequest(FormAuthenticator.java:260)
	at org.eclipse.jetty.security.authentication.DeferredAuthentication.authenticate(DeferredAuthentication.java:68)
	at org.eclipse.jetty.server.Request.getUserPrincipal(Request.java:1479)
	at org.eclipse.jetty.server.Request.getRemoteUser(Request.java:1087)
	at winstone.accesslog.SimpleAccessLogger.log(SimpleAccessLogger.java:89)
	at org.eclipse.jetty.server.handler.RequestLogHandler.handle(RequestLogHandler.java:113)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
	at org.eclipse.jetty.server.Server.handle(Server.java:499)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:311)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:257)
	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:544)
	at winstone.BoundedExecutorService$1.run(BoundedExecutorService.java:77)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
```

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
CC @reviewbybees

File: src/java/winstone/accesslog/SimpleAccessLogger.java [hunk 1]
```diff
@@ -86,6 +87,17 @@ public void log(Request request, Response response) {
...
         }
+
+        // mimic https://github.com/eclipse/jetty.project/blob/jetty-9.4.0.v20161208/jetty-server/src/main/java/org/eclipse/jetty/server/AbstractNCSARequestLog.java#L130
+        Authentication authentication = request.getAuthentication();
+        String remoteUser;
+        if (authentication instanceof Authentication.User) {
+            Authentication.User user = (Authentication.User) authentication;
+            remoteUser = user.getUserIdentity().getUserPrincipal().getName();
+        } else {
+            remoteUser = null;
```
Comment (batmat) [inline src/java/winstone/accesslog/SimpleAccessLogger.java]: NIT: I'm wondering if logging then "NO USER" or something wouldn't be better. Maybe not and null is clearer in that regard.

**Comment by cyrille-leclerc** [comment]
Build failed due to a platform error, not to a regression

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:2.4:compile (default-compile) on project winstone: Execution default-compile of goal org.apache.maven.plugins:maven-compiler-plugin:2.4:compile failed: Plugin org.apache.maven.plugins:maven-compiler-plugin:2.4 or one of its dependencies could not be resolved: Could not transfer artifact org.apache.maven:maven-core:jar:2.0.9 from/to central (https://repo.maven.apache.org/maven2): GET request of: org/apache/maven/maven-core/2.0.9/maven-core-2.0.9.jar from central failed: Connection reset -> [Help 1]
```

**Comment by oleg-nenashev** [comment]
` Could not transfer artifact org.apache.maven:maven-core:jar:2.0.9 from/to central (https://repo.maven.apache.org/maven2): GET request of: org/apache/maven/maven-core/2.0.9/maven-core-2.0.9.jar` likely says that it's a good time to update Winstone a bit. I think it's safe to update Java3 to Java6 dependency at this point

File: src/java/winstone/accesslog/SimpleAccessLogger.java (hunk 1 above)
Comment (oleg-nenashev) [inline src/java/winstone/accesslog/SimpleAccessLogger.java]: `nvl()` below replaces null by dash. I feel it's a right behavior here

**Comment by cyrille-leclerc** [comment]
@oleg-nenashev regarding you comment
> Though I would rather expect [Request#getRemoteUser()](http://download.eclipse.org/jetty/stable-9/apidocs/org/eclipse/jetty/server/Request.html#getRemoteUser--) to handle it.

My understanding was that [org.eclipse.jetty.server.Request#getRemoteUser()](http://download.eclipse.org/jetty/stable-9/apidocs/org/eclipse/jetty/server/Request.html#getRemoteUser--) is the implementation of [javax.servlet.http.HttpServletRequest#getRemoteUser()](http://docs.oracle.com/javaee/6/api/javax/servlet/http/HttpServletRequest.html#getRemoteUser()) and is designed to be invoked in servlets and servlet wrappers but is not designed to be invoked in Jetty's internal component [org.eclipse.jetty.server.RequestLog](http://download.eclipse.org/jetty/stable-9/apidocs/org/eclipse/jetty/server/RequestLog.html).

I assumed that this was the reason why [org.eclipse.jetty.server.Request#getRemoteUser()](http://download.eclipse.org/jetty/stable-9/apidocs/org/eclipse/jetty/server/Request.html#getRemoteUser--)  was misbehaving in [org.eclipse.jetty.server.RequestLog#log](http://download.eclipse.org/jetty/stable-9/apidocs/org/eclipse/jetty/server/RequestLog.html#log-org.eclipse.jetty.server.Request-org.eclipse.jetty.server.Response-).

**Comment by stephenc** [comment]
:bee:

**Pull request merged**